File: server/src/api/schedules.ts
import { Hono } from "hono";
import { z } from "zod";
import { zValidator } from "@hono/zod-validator";
import {
  createSchedule,
  deleteSchedule,
  getSchedule,
  listSchedules,
  updateSchedule,
} from "../services/schedules.js";

export const schedulesRouter = new Hono();

// --- Schemas ---
const createScheduleSchema = z.object({
  name: z.string().min(1),
  description: z.string().optional(),
  cron_expression: z.string().min(1),
  task_prompt: z.string().min(1),
  origin_session_id: z.string().optional(),
  max_runs: z.number().int().positive().optional(),
  metadata: z.record(z.unknown()).optional(),
});

const updateScheduleSchema = z.object({
  name: z.string().min(1).optional(),
  description: z.string().optional(),
  cron_expression: z.string().min(1).optional(),
  task_prompt: z.string().min(1).optional(),
  enabled: z.boolean().optional(),
  max_runs: z.number().int().positive().nullable().optional(),
});

// --- GET /v1/schedules ---
schedulesRouter.get("/", async (c) => {
  const limit = parseInt(c.req.query("limit") ?? "20", 10);
  const offset = parseInt(c.req.query("offset") ?? "0", 10);
  return c.json(await listSchedules(limit, offset));
});

// --- POST /v1/schedules ---
schedulesRouter.post(
  "/",
  zValidator("json", createScheduleSchema),
  async (c) => {
    const body = c.req.valid("json");
    try {
      const schedule = await createSchedule({
        name: body.name,
        description: body.description,
        cronExpression: body.cron_expression,
        taskPrompt: body.task_prompt,
        originSessionId: body.origin_session_id,
        maxRuns: body.max_runs,
        metadata: body.metadata,
      });
      return c.json(schedule, 201);
    } catch (err: any) {
      return c.json({ error: err.message }, 400);
    }
  },
);

// --- GET /v1/schedules/:id ---
schedulesRouter.get("/:id", async (c) => {
  const id = String(c.req.param("id"));
  const schedule = await getSchedule(id);
  if (!schedule) return c.json({ error: "Schedule not found" }, 404);
  return c.json(schedule);
});

// --- PATCH /v1/schedules/:id ---
schedulesRouter.patch(
  "/:id",
  zValidator("json", updateScheduleSchema),
  async (c) => {
    const id = String(c.req.param("id"));
    const body = c.req.valid("json");
    try {
      const schedule = await updateSchedule(id, {
        name: body.name,
        description: body.description,
        cronExpression: body.cron_expression,
        taskPrompt: body.task_prompt,
        enabled: body.enabled,
        maxRuns: body.max_runs,
      });
      if (!schedule) return c.json({ error: "Schedule not found" }, 404);
      return c.json(schedule);
    } catch (err: any) {
      return c.json({ error: err.message }, 400);
    }
  },
);

// --- DELETE /v1/schedules/:id ---
schedulesRouter.delete("/:id", async (c) => {
  const id = String(c.req.param("id"));
  const deleted = await deleteSchedule(id);
  return c.json({ deleted, id });
});

File: sdk/python/kraken/schedules.py
"""Schedule management operations."""

from __future__ import annotations

from typing import Any, List  # noqa: UP035

from kraken._transport import Transport
from kraken.models import Schedule


class Schedules:
    """Manage recurring and one-time scheduled tasks."""

    def __init__(self, transport: Transport) -> None:
        self._t = transport

    def list(self, *, limit: int = 20, offset: int = 0) -> List[Schedule]:  # noqa: UP006
        """List schedules."""
        data = self._t.get("/v1/schedules", params={"limit": limit, "offset": offset})
        return [Schedule.model_validate(s) for s in data.get("schedules", [])]

    def create(
        self,
        name: str,
        task_prompt: str,
        cron_expression: str,
        *,
        description: str | None = None,
        origin_session_id: str | None = None,
        max_runs: int | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Schedule:
        """Create a recurring schedule."""
        payload: dict[str, Any] = {
            "name": name,
            "task_prompt": task_prompt,
            "cron_expression": cron_expression,
        }
        if description is not None:
            payload["description"] = description
        if origin_session_id is not None:
            payload["origin_session_id"] = origin_session_id
        if max_runs is not None:
            payload["max_runs"] = max_runs
        if metadata is not None:
            payload["metadata"] = metadata
        data = self._t.post("/v1/schedules", json=payload)
        return Schedule.model_validate(data)

    def get(self, schedule_id: str) -> Schedule:
        """Get a schedule by ID."""
        data = self._t.get(f"/v1/schedules/{schedule_id}")
        return Schedule.model_validate(data)

    def update(
        self,
        schedule_id: str,
        *,
        name: str | None = None,
        description: str | None = None,
        cron_expression: str | None = None,
        task_prompt: str | None = None,
        enabled: bool | None = None,
        max_runs: int | None = None,
    ) -> Schedule:
        """Update a schedule."""
        payload: dict[str, Any] = {}
        if name is not None:
            payload["name"] = name
        if description is not None:
            payload["description"] = description
        if cron_expression is not None:
            payload["cron_expression"] = cron_expression
        if task_prompt is not None:
            payload["task_prompt"] = task_prompt
        if enabled is not None:
            payload["enabled"] = enabled
        if max_runs is not None:
            payload["max_runs"] = max_runs
        data = self._t.patch(f"/v1/schedules/{schedule_id}", json=payload)
        return Schedule.model_validate(data)

    def delete(self, schedule_id: str) -> None:
        """Delete a schedule."""
        self._t.delete(f"/v1/schedules/{schedule_id}")
